File: packages/schemas/src/memory-page.ts
import { z } from "zod";
import type { Actor, JsonValue } from "./index.js";

const nonEmptyString = z.string().trim().min(1);
const dateTime = z.string().datetime({ offset: true });

const memoryPageSensitivitySchema = z.enum([
  "ordinary",
  "personal",
  "sensitive",
  "restricted",
]);

const memoryPageHorizonSchema = z.enum([
  "durable",
  "project",
  "temporary",
]);

export const memoryPageStatusSchema = z.enum([
  "candidate",
  "confirmed",
  "rejected",
  "superseded",
  "expired",
]);
export type MemoryPageStatus = z.infer<typeof memoryPageStatusSchema>;
export const MEMORY_PAGE_STATUSES = memoryPageStatusSchema.options;

export const memoryPageOriginSchema = z.enum([
  "manual",
  "extracted",
  "imported",
  "compatibility",
]);
export type MemoryPageOrigin = z.infer<typeof memoryPageOriginSchema>;

export const memoryPageSourceRefSchema = z
  .object({
    sourceId: nonEmptyString,
    evidence: nonEmptyString.optional(),
    turnIds: z
      .array(nonEmptyString)
      .refine(
        (turnIds) => new Set(turnIds).size === turnIds.length,
        "turnIds must be unique",
      )
      .optional(),
  })
  .strict();
export type MemoryPageSourceRef = z.infer<typeof memoryPageSourceRefSchema>;

export const memoryPageSchema = z
  .object({
    id: nonEmptyString,
    subject: nonEmptyString,
    title: nonEmptyString,
    summary: nonEmptyString.optional(),
    body: nonEmptyString,
    category: nonEmptyString.optional(),
    tags: z.array(nonEmptyString).refine(
      (tags) => new Set(tags).size === tags.length,
      "tags must be unique",
    ),
    status: memoryPageStatusSchema,
    sensitivity: memoryPageSensitivitySchema,
    horizon: memoryPageHorizonSchema,
    origin: memoryPageOriginSchema,
    sourceRefs: z.array(memoryPageSourceRefSchema).min(1),
    annotationIds: z.array(nonEmptyString).refine(
      (ids) => new Set(ids).size === ids.length,
      "annotationIds must be unique",
    ),
    validFrom: dateTime.optional(),
    validUntil: dateTime.optional(),
    supersedes: z.array(nonEmptyString).refine(
      (ids) => new Set(ids).size === ids.length,
      "supersedes must be unique",
    ),
    revision: z.number().int().positive(),
    createdAt: dateTime,
    updatedAt: dateTime,
  })
  .strict()
  .superRefine((page, context) => {
    if (
      page.validFrom !== undefined &&
      page.validUntil !== undefined &&
      Date.parse(page.validUntil) < Date.parse(page.validFrom)
    ) {
      context.addIssue({
        code: "custom",
        path: ["validUntil"],
        message: "validUntil cannot be before validFrom",
      });
    }

    if (page.supersedes.includes(page.id)) {
      context.addIssue({
        code: "custom",
        path: ["supersedes"],
        message: "a Memory Page cannot supersede itself",
      });
    }
  });
export type MemoryPage = z.infer<typeof memoryPageSchema>;

export type MemoryPageEventType =
  | "memory.proposed"
  | "memory.confirmed"
  | "memory.edited"
  | "memory.rejected"
  | "memory.superseded"
  | "memory.expired";

export interface MemoryPageEvent {
  id: string;
  type: MemoryPageEventType;
  entityType: "memory";
  entityId: string;
  occurredAt: string;
  actor: Actor;
  data?: Record<string, JsonValue>;
}

export interface MemoryPageTransition {
  page: MemoryPage;
  event: MemoryPageEvent;
}

export function validateMemoryPage(value: unknown): asserts value is MemoryPage {
  const result = memoryPageSchema.safeParse(value);
  if (!result.success) {
    throw new Error(
      `Invalid TOPO Memory Page: ${result.error.issues
        .map((issue) => {
          const path = issue.path.length > 0 ? `${issue.path.join(".")}: ` : "";
          return `${path}${issue.message}`;
        })
        .join("; ")}`,
    );
  }
}

File: crates/topo-contracts/src/memory_page.rs
use crate::{Actor, EpistemicType, MemoryHorizon, Sensitivity};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::collections::BTreeMap;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum MemoryPageStatus {
    Candidate,
    Confirmed,
    Rejected,
    Superseded,
    Expired,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum MemoryPageOrigin {
    Manual,
    Extracted,
    Imported,
    Compatibility,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MemoryPageSourceRef {
    pub source_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub evidence: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub turn_ids: Option<Vec<String>>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MemoryPage {
    pub id: String,
    pub subject: String,
    pub title: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub summary: Option<String>,
    pub body: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub category: Option<String>,
    pub tags: Vec<String>,
    pub status: MemoryPageStatus,
    pub sensitivity: Sensitivity,
    pub horizon: MemoryHorizon,
    pub origin: MemoryPageOrigin,
    pub source_refs: Vec<MemoryPageSourceRef>,
    pub annotation_ids: Vec<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub valid_from: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub valid_until: Option<String>,
    pub supersedes: Vec<String>,
    pub revision: u64,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MemoryPageAnnotationProposal {
    pub key: String,
    pub value: Value,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub category: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tags: Option<Vec<String>>,
    pub epistemic_type: EpistemicType,
    pub confidence: f64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sensitivity: Option<Sensitivity>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ExtractedMemoryPageProposal {
    pub title: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub summary: Option<String>,
    pub body: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub category: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tags: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sensitivity: Option<Sensitivity>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub horizon: Option<MemoryHorizon>,
    pub evidence_turn_ids: Vec<String>,
    pub evidence: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub valid_from: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub valid_until: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub annotations: Option<Vec<MemoryPageAnnotationProposal>>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub enum MemoryPageEventType {
    #[serde(rename = "memory.proposed")]
    Proposed,
    #[serde(rename = "memory.confirmed")]
    Confirmed,
    #[serde(rename = "memory.edited")]
    Edited,
    #[serde(rename = "memory.rejected")]
    Rejected,
    #[serde(rename = "memory.superseded")]
    Superseded,
    #[serde(rename = "memory.expired")]
    Expired,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MemoryPageEvent {
    pub id: String,
    #[serde(rename = "type")]
    pub event_type: MemoryPageEventType,
    pub entity_type: String,
    pub entity_id: String,
    pub occurred_at: String,
    pub actor: Actor,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub data: Option<BTreeMap<String, Value>>,
}
